File: px-services/src/main/java/com/mercadopago/android/px/model/internal/Modal.java
package com.mercadopago.android.px.model.internal;

import android.os.Parcel;
import android.os.Parcelable;

public final class Modal implements Parcelable {

    private final Text title;
    private final Text description;
    private final Action mainButton;
    private final Action secondaryButton;
    private final String iconUrl;

    public static final Creator<Modal> CREATOR = new Creator<Modal>() {
        @Override
        public Modal createFromParcel(final Parcel in) {
            return new Modal(in);
        }

        @Override
        public Modal[] newArray(final int size) {
            return new Modal[size];
        }
    };

    @SuppressWarnings({ "WeakerAccess", "ProtectedMemberInFinalClass" })
    protected Modal(final Parcel in) {
        title = in.readParcelable(Text.class.getClassLoader());
        description = in.readParcelable(Text.class.getClassLoader());
        mainButton = in.readParcelable(Action.class.getClassLoader());
        secondaryButton = in.readParcelable(Action.class.getClassLoader());
        iconUrl = in.readString();
    }

    @Override
    public void writeToParcel(final Parcel dest, final int flags) {
        dest.writeParcelable(title, flags);
        dest.writeParcelable(description, flags);
        dest.writeParcelable(mainButton, flags);
        dest.writeParcelable(secondaryButton, flags);
        dest.writeString(iconUrl);
    }

    public Text getTitle() {
        return title;
    }

    public Text getDescription() {
        return description;
    }

    public Action getMainButton() {
        return mainButton;
    }

    public Action getSecondaryButton() {
        return secondaryButton;
    }

    public String getIconUrl() {
        return iconUrl;
    }

    @Override
    public int describeContents() {
        return 0;
    }
}

File: px-services/src/main/java/com/mercadopago/android/px/model/internal/Modal.kt
package com.mercadopago.android.px.model.internal

import android.os.Parcel
import com.google.gson.annotations.SerializedName
import com.mercadopago.android.px.internal.util.KParcelable
import com.mercadopago.android.px.internal.util.parcelableCreator

data class Modal(
    val title: Text,
    val description: Text,
    val mainButton: Action,
    val secondaryButton: Action?,
    @SerializedName("icon_url") val imageUrl: String?) : KParcelable {

    constructor(parcel: Parcel) : this(
        parcel.readParcelable(Text::class.java.classLoader)!!,
        parcel.readParcelable(Text::class.java.classLoader)!!,
        parcel.readParcelable(Action::class.java.classLoader)!!,
        parcel.readParcelable(Action::class.java.classLoader),
        parcel.readString())

    override fun writeToParcel(parcel: Parcel, flags: Int) = with(parcel) {
        writeParcelable(title, flags)
        writeParcelable(description, flags)
        writeParcelable(mainButton, flags)
        writeParcelable(secondaryButton, flags)
        writeString(imageUrl)
    }

    companion object {
        @JvmField
        val CREATOR = parcelableCreator(::Modal)
    }
}